### ISSUE TITLE
No HEVC hardware decode for 25 fps videos

### ISSUE THREAD

**Issue by edwrap** [mpv-android/mpv-android#331]
Device: Samsung Galaxy Tab S6

Issue: HEVC Main10 hardware decode works fine for 24 fps content (1080p/4K), but does not activate for 25 fps (1080p/4K). 1080p can be played fine in software, but 4K is unwatchable. VLC hardware decoding seems to work for 4K 25 fps, so it doesn’t appear to be a device capability issue.

**Comment by sfan5** [comment]
Please provide [logcat](https://developer.android.com/studio/command-line/logcat) output.

**Comment by edwrap** [comment]
Attached is full mpv logcat output for two videos, one at 1080p25, the other at 4k24, both HEVC with HDR. 

[1080p25.txt](https://github.com/mpv-android/mpv-android/files/5548571/1080p25.txt)
[4k24.txt](https://github.com/mpv-android/mpv-android/files/5548572/4k24.txt)

**Comment by edwrap** [comment]
Let me know if video samples would be helpful. I’ve since come across an 1080p H264 50fps file that also doesn’t use hardware decoding, though software decoding is enough. I’ve also confirmed that 4k 25fps videos play smoothly in the built-in player. I would use that, but it doesn’t seem to support streaming from network sources. As for VLC, it doesn’t appear to be doing HDR, though I could be mistaken.

**Comment by tarkov2213** [comment]
mpv fails to toggle HW on a 23.976 FPS 2160p HDR HEVC video on the essential phone or the fire TV stick. HW works fine on both devices on MX player and VLC. In the fire TV case HDR is working on both those players and on the phone there is no HDR so it perform tone mapping.

**Comment by sfan5** [comment]
> Let me know if video samples would be helpful.

Definitely, the issue here isn't the framerate but some other circumstance.

Specifically this line catches my eye: `[ffmpeg:warn] No parameter sets in the extradata`
FFmpeg's decoder expects to extract some data [from the extradata](https://github.com/FFmpeg/FFmpeg/blob/50a92107393113cf0c699877536978adffcc9004/libavcodec/mediacodecdec.c#L182-L269) (though it *should* have logged an error upon not finding it).

Since mpv is not involved at this point you can also try reporting this on the [ffmpeg bugtracker](http://ffmpeg.org/bugreports.html) to directly reach someone more familiar with this code.

> worth pointing out that even when it does play with HW it has some dropped frames

Some loss of performance is to be expected since mpv always uses OpenGL rendering and (currently?) can't use zero-copy hardware decoding on Android.

**Comment by edwrap** [comment]
Don't have the 4K24HDR file anymore, but hopefully these are enough:

https://user-images.githubusercontent.com/59624432/102697140-52a9a600-41e8-11eb-9c49-e11f066ed585.mp4


https://user-images.githubusercontent.com/59624432/102697141-53dad300-41e8-11eb-8016-ab90242e4b38.mp4

**Comment by sfan5** [comment]
Right, here are the warnings when playing that:
<pre>
12-21 22:22:44.447   432   489 V mpv     : [vd:v] Using underlying hw-decoder 'h264_mediacodec'
12-21 22:22:44.447   432   489 V mpv     : [ffmpeg:warn] h264_mp4toannexb: Warning: SPS NALU missing or invalid. The resulting stream may not play.
12-21 22:22:44.447   432   489 V mpv     : [ffmpeg:warn] h264_mp4toannexb: Warning: PPS NALU missing or invalid. The resulting stream may not play.
12-21 22:22:44.447   432   489 V mpv     : [ffmpeg/video:error] h264_mediacodec: Could not extract PPS/SPS from extradata
</pre>

but if you patch the function not to return an error it plays anyway:
<pre>
12-21 22:22:44.731   432   489 V mpv     : [ffmpeg/video:v] h264_mediacodec: Output crop parameters top=0 bottom=1079 left=0 right=1919, resulting dimensions width=1920 height=1080
12-21 22:22:44.731   432   489 V mpv     : [ffmpeg/video:v] h264_mediacodec: MediaCodec started successfully: codec = OMX.Exynos.avc.dec, ret = 0
[...]
12-21 22:22:44.799   432   489 V mpv     : [ffmpeg/video:v] h264_mediacodec: Output MediaFormat changed to {crop-right=1919, vendor.rtc-ext-dec-caps-vt-driver-version.number=0, vendor.rtc-ext-dec-low-latency.enable=0, color-format=21, slice-height=1088, image-data=java.nio.HeapByteBuffer[pos=0 lim=104 cap=104], mime=video/raw, vendor.sec-ext-dec-compressed-color-format=0, stride=1920, color-standard=1, color-transfer=3, crop-bottom=1079, crop-left=0, width=1920, crop-top=0, height=1088}
</pre>

The HEVC/HDR sample doesn't play on my device either way, so can't test that.

**Comment by edwrap** [comment]
Thanks for looking into this! Sorry, those errors look to be from remuxing to mp4? GitHub only allowed video upload as mp4, and I thought that might be preferable to a zip file. I tested the files in mpv and they still showed SW decode, so I figured that was okay, but I can slice some cleaner mkv samples in the morning if that would be helpful. Not sure from your post if you have enough information to possibly fix this yet
